iroh/src/util.rs

```diff
@@ -6,17 +6,44 @@ use std::{
     task::{Context, Poll},
 };
 
-/// Resolves to pending if the inner is `None`.
+/// A future which may not be present.
+///
+/// This is a single type which may optionally contain a future.  If there is no inner
+/// future polling will always return [`Poll::Pending`].
+///
+/// The [`Default`] impl will create a [`MaybeFuture`] without an inner.
 #[derive(Debug)]
 pub(crate) struct MaybeFuture<T> {
     /// Future to be polled.
     pub inner: Option<T>,
 }
 
+impl<T> MaybeFuture<T> {
+    /// Creates a [`MaybeFuture`] without an inner future.
+    pub(crate) fn none() -> Self {
```

Comment:
I've leant into the `Option` sematics and gone with `::none()` constructor and `.is_some()` and `.is_none()`.  This seems to match the naming which is almost `OptionalFuture`.

The alternative is to lean into collection APIs: `::empty()` as constructor and `.is_empty()` to check (I guess `.len()` would be it's counterpart but the clippy convention is to use `!m.is_empty()`).

Reply:
I like the Option version